trie-db/src/lookup.rs

```diff
@@ -655,4 +668,175 @@ where
+						let common_prefix_len = partial.common_prefix(&slice);
+
+						let start_matches = if partial.len() <= slice.len() {
+							// Extension slice can be longer than remainder of the provided key
+							// (descendent), ensure the extension slice starts with the remainder
+							// of the provided key.
+							//
+							// This effectively returns the hash of the `Node::Branch` at the next
+							// iteration.
+							common_prefix_len == partial.len()
+						} else {
+							// Remainder of the provided key is longer than the extension slice,
+							// must advance the node iteration if and only if keys share
+							// a common prefix.
+							common_prefix_len == slice.len()
+						};
+
+						if start_matches {
+							// Empties the partial key if the extension slice is longer.
+							partial = partial.mid(common_prefix_len);
+							key_nibbles += common_prefix_len;
+							item
+						} else {
+							self.record(|| TrieAccess::NonExisting { full_key });
+
+							return Ok(None)
+						}
+					},
+					Node::Branch(children, value) =>
+						if partial.is_empty() {
+							if value.is_none() {
+								self.record(|| TrieAccess::NonExisting { full_key });
+							}
+
+							return Ok(Some(hash))
+						} else {
+							match children[partial.at(0) as usize] {
+								Some(x) => {
+									partial = partial.mid(1);
+									key_nibbles += 1;
+									x
+								},
+								None => {
+									self.record(|| TrieAccess::NonExisting { full_key });
+
+									return Ok(None)
+								},
+							}
+						},
+					Node::NibbledBranch(slice, children, value) => {
+						let common_prefix_len = partial.common_prefix(&slice);
+						// Not enough remainder key to continue the search.
+						if partial.len() < slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+
+							// Branch slice starts with the remainder key, there's nothing to
+							// advance.
+							if common_prefix_len == partial.len() {
+								return Ok(Some(hash))
+							} else {
+								return Ok(None)
+							}
+						}
+
+						// Partial key is longer or equal than the branch slice.
+						// Ensure partial key starts with the branch slice.
+						if common_prefix_len != slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+							return Ok(None)
+						}
+
+						if partial.len() == slice.len() {
+							if value.is_none() {
+								self.record(|| TrieAccess::NonExisting { full_key });
+							}
+
+							return Ok(Some(hash))
+						} else {
+							match children[partial.at(slice.len()) as usize] {
+								Some(x) => {
+									partial = partial.mid(slice.len() + 1);
+									key_nibbles += slice.len() + 1;
+									x
+								},
+								None => {
+									self.record(|| TrieAccess::NonExisting { full_key });
+
+									return Ok(None)
+								},
+							}
+						}
+					},
+					Node::Empty => {
+						self.record(|| TrieAccess::NonExisting { full_key });
+
+						return Ok(Some(hash))
```

Comment:
I would return Ok(None) (rational being that the queried key is not contained in the trie).